src/main/scala/analysis/RegionInjector.scala

```diff
@@ -0,0 +1,321 @@
+      case SignExtend(extension, body) =>
+        SignExtend(extension, eval(body, cmd))
+      case UnaryExpr(op, arg) =>
+        UnaryExpr(op, eval(arg, cmd))
+      case BinaryExpr(op, arg1, arg2) =>
+        BinaryExpr(op, eval(arg1, cmd), eval(arg2, cmd))
+      case MemoryLoad(mem, index, endian, size) =>
+        // TODO: index should be replaced region
+        MemoryLoad(renameMemory(mem, index, cmd), eval(index, cmd), endian, size)
+      case variable: Variable => variable // ignore variables
+  }
+
+  def renameMemory(mem: Memory, expr: Expr, cmd : Command): Memory = {
+    val regions = exprToRegion(eval(expr, cmd), cmd)
+    if (regions.size == 1) {
+      Logger.warn(s"Mem CMD is: ${cmd}")
+      Logger.warn(s"Region found for mem: ${regions.head}")
+      regions.head match {
+        case stackRegion: StackRegion =>
+          return StackMemory(stackRegion.regionIdentifier, mem.addressSize, mem.valueSize)
+        case dataRegion: DataRegion =>
+          return SharedMemory(dataRegion.regionIdentifier, mem.addressSize, mem.valueSize)
+        case _ =>
+      }
+    } else if (regions.size > 1) {
+      Logger.warn(s"Mem CMD is: ${cmd}")
+      Logger.warn(s"Multiple regions found for mem: ${regions}")
+      mmm.mergeRegions(regions) match {
+        case stackRegion: StackRegion =>
+          return StackMemory(stackRegion.regionIdentifier, mem.addressSize, mem.valueSize)
+        case dataRegion: DataRegion =>
+          return SharedMemory(dataRegion.regionIdentifier, mem.addressSize, mem.valueSize)
+        case _ =>
+      }
+    } else {
+      Logger.warn(s"Mem CMD is: ${cmd}")
+      Logger.warn(s"No region found for mem")
+    }
+    mem
+  }
+
+  /** Transfer function for state lattice elements.
+   */
+  def localTransfer(n: CFGPosition): Unit = n match {
+    case cmd: Command =>
+      cmd match
+        case statement: Statement => statement match
+          case assign: Assign =>
+            assign.rhs = eval(assign.rhs, cmd)
+          case mAssign: MemoryAssign =>
+            mAssign.mem = renameMemory(mAssign.mem, mAssign.index, cmd)
+            mAssign.index = eval(mAssign.index, cmd)
+            mAssign.value = eval(mAssign.value, cmd)
+          case nop: NOP => // ignore NOP
+          case assert: Assert =>
+            assert.body = eval(assert.body, cmd)
+          case assume: Assume =>
+            assume.body = eval(assume.body, cmd)
+        case jump: Jump => jump match
+          case to: GoTo => // ignore GoTo
+          case call: Call => call match
+            case call: DirectCall => // ignore DirectCall
+            case call: IndirectCall => // ignore IndirectCall
+    case _ => // ignore other kinds of nodes
+  }
+
+  def transformMemorySections(memorySegment: ArrayBuffer[MemorySection]): ArrayBuffer[MemorySection] = {
+    val newArrayBuffer = ArrayBuffer.empty[MemorySection]
+    for (elem <- memorySegment) {
+      elem match {
+        case mem: MemorySection =>
```

Comment:
there isn't a need to match this, it's always going to be a MemorySection?

Reply:
You are right. I think this is left over from the changes to stmts and exprs.